spatialos-sdk/src/worker/alpha.rs

```diff
@@ -0,0 +1,465 @@
+                    .collect::<Vec<String>>();
+
+            LoginTokenDetails {
+                deployment_id: cstr_to_string(token_details.deployment_id),
+                deployment_name: cstr_to_string(token_details.deployment_name),
+                tags,
+                login_token: cstr_to_string(token_details.login_token),
+            }
+        }
+    }
+}
+
+pub struct LoginTokensFuture {
+    internal: *mut Worker_Alpha_LoginTokensResponseFuture,
+    consumed: bool,
+}
+
+impl LoginTokensFuture {
+    fn new(ptr: *mut Worker_Alpha_LoginTokensResponseFuture) -> Self {
+        LoginTokensFuture {
+            internal: ptr,
+            consumed: false,
+        }
+    }
+
+    extern "C" fn callback_handler(
+        user_data: *mut ::std::os::raw::c_void,
+        response: *const Worker_Alpha_LoginTokensResponse,
+    ) {
+        assert!(!response.is_null());
+        unsafe {
+            let response = *response;
+            let data = &mut *(user_data as *mut Option<Result<LoginTokensResponse, String>>);
+            if !response.error.is_null() {
+                let err = cstr_to_string(response.error);
+                *data = Some(Err(err));
+                return;
+            }
+
+            let response = LoginTokensResponse::from_worker_sdk(&response);
+            *data = Some(Ok(response));
+        }
+    }
+
+    fn get_future(&self, timeout: Option<u32>) -> Option<Result<LoginTokensResponse, String>> {
+        assert!(!self.internal.is_null());
+        let mut data: Option<Result<LoginTokensResponse, String>> = None;
+        unsafe {
+            Worker_Alpha_LoginTokensResponseFuture_Get(
+                self.internal,
+                timeout.map_or(::std::ptr::null(), |value| &value),
+                &mut data as *mut _ as *mut ::std::os::raw::c_void,
+                Some(LoginTokensFuture::callback_handler),
+            );
+
+            data
+        }
+    }
+}
+
+impl Future for LoginTokensFuture {
+    type Item = LoginTokensResponse;
+    type Error = String;
+
+    fn poll(&mut self) -> Result<Async<<Self as Future>::Item>, <Self as Future>::Error> {
+        if self.consumed {
+            return Err("LoginTokensFuture has already been consumed".to_owned());
+        }
+
+        self.get_future(Some(0))
+            .map_or(Ok(Async::NotReady), |result| {
+                self.consumed = true;
+                result.map(Async::Ready)
+            })
+    }
+
+    fn wait(self) -> Result<<Self as Future>::Item, <Self as Future>::Error>
+        where
+            Self: Sized,
+    {
+        if self.consumed {
+            return Err("LoginTokensFuture has already been consumed".to_owned());
+        }
+
+        self.get_future(None)
+            .expect("Blocking call to Worker_Alpha_LoginTokensFuture_Get did not trigger callback")
+    }
+}
+
+impl Drop for LoginTokensFuture {
+    fn drop(&mut self) {
+        if !self.internal.is_null() {
```

Comment:
I think the null check here is unnecessary? The pointer doesn't change over the lifetime of the object, and it only ever makes sense to construct the future with a non-null pointer, so it would probably make more sense to assert that the pointer isn't null in `LoginTokensFuture::new()`

Reply:
Errr... yes you are correct!